nemo/core/classes/modelPT.py

```diff
@@ -1831,24 +1864,50 @@ def on_train_batch_start(self, batch: Any, batch_idx: int, unused: int = 0) -> O
         # nsys profiling
         if self.device.type == 'cuda':
             if hasattr(self, '_nsys_profile_enabled'):
                 if self._nsys_profile_enabled and not self._nsys_profile_started:
-                    if batch_idx >= self._nsys_profile_start_step and get_rank() in self._nsys_profile_ranks:
+                    if self._real_batch_idx >= self._nsys_profile_start_step and get_rank() in self._nsys_profile_ranks:
                         logging.info("====== Start nsys profiling ======")
                         torch.cuda.cudart().cudaProfilerStart()
                         if self._nsys_profile_gen_shape:
                             torch.autograd.profiler.emit_nvtx(record_shapes=True).__enter__()
                         self._nsys_profile_started = True
 
             if hasattr(self, '_memory_profile_enabled'):
+                if self._memory_profile_enabled:
+                    if self._memory_profile_weight_enabled and (self._real_batch_idx == 0) and (get_rank() == self._memory_profile_rank): # before batch-0 start
+                        try:
+                            logging.info(f"====== Rank[{self._memory_profile_rank}], Batch[{batch_idx}]. End CUDA memory profiling: Weight. Snapshot saved in {self._memory_profile_snapshot_file_weight} ======")
+                            torch.cuda.memory._dump_snapshot(f"{self._memory_profile_snapshot_file_weight}")
+                            # torch.cuda.memory._record_memory_history(enabled=None)
+                        except Exception as e:
+                            logging.error(f"Failed to capture memory snapshot {e}")
+                            return
+
+                        # Call the analysis function
+                        if self._memory_profile_analysis_enabled:
+                            if not os.path.exists(self._memory_profile_analysis_path):
+                                os.makedirs(self._memory_profile_analysis_path)
+                            if os.path.exists(self._memory_profile_snapshot_file_weight):
+                                logging.info(f"====== Memory Profile Analysis: Weight ======")
+                                peak_memory_analysis(self._memory_profile_snapshot_file_weight, self._memory_profile_analysis_path, 'weight', self._memory_profile_rank)
+                            else:
+                                raise Exception(f"Snapshot file not found: {self._memory_profile_snapshot_file_weight}")    
+
+
                 if self._memory_profile_enabled and not self._memory_profile_started:
-                    if batch_idx >= self._memory_profile_start_step and get_rank() == self._memory_profile_rank:
-                        logging.info("====== Start CUDA memory profiling ======")
-                        torch.cuda.memory._record_memory_history(max_entries=100000)
+                    if self._real_batch_idx == self._memory_profile_start_step and get_rank() == self._memory_profile_rank:
```

Comment:
why is it `self._real_batch_idx == self._memory_profile_start_step ` instead of `self._real_batch_idx >= self._memory_profile_start_step ` ?